browser/ui/webui/ai_chat/ai_chat_ui_page_handler.h

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_BROWSER_UI_WEBUI_AI_CHAT_AI_CHAT_UI_PAGE_HANDLER_H_
+#define BRAVE_BROWSER_UI_WEBUI_AI_CHAT_AI_CHAT_UI_PAGE_HANDLER_H_
+
+#include <stdint.h>
+
+#include <memory>
+#include <string>
+
+#include "base/scoped_observation.h"
+#include "brave/browser/ai_chat/ai_chat.mojom.h"
+#include "brave/browser/ai_chat/ai_chat_tab_helper.h"
+#include "chrome/browser/ui/tabs/tab_strip_model_observer.h"
+#include "mojo/public/cpp/bindings/pending_receiver.h"
+#include "mojo/public/cpp/bindings/pending_remote.h"
+#include "mojo/public/cpp/bindings/receiver.h"
+#include "mojo/public/cpp/bindings/remote.h"
+
+class TabStripModel;
+
+class AIChatUIPageHandler : public ai_chat::mojom::PageHandler,
+                            public TabStripModelObserver,
+                            public AIChatTabHelper::Observer {
+ public:
+  AIChatUIPageHandler(
+      TabStripModel* tab_strip_model,
+      Profile* profile,
+      mojo::PendingReceiver<ai_chat::mojom::PageHandler> receiver);
+
+  AIChatUIPageHandler(const AIChatUIPageHandler&) = delete;
+  AIChatUIPageHandler& operator=(const AIChatUIPageHandler&) = delete;
+
+  ~AIChatUIPageHandler() override;
+
+  // ai_chat::mojom::PageHandler:
+  void SetClientPage(
+      mojo::PendingRemote<ai_chat::mojom::ChatUIPage> page) override;
+  void SubmitHumanConversationEntry(const std::string& input) override;
+  void GetConversationHistory(GetConversationHistoryCallback callback) override;
+  void RequestSummary() override;
```

Comment:
Why are we using an observer to respond to this instead of a callback?

Reply:
so there is a larger problem here because of the async network/renderer calls and `SetRequestInProgress` which can go wrong in a lot of different ways if another call comes in before the previous one has finished when navigating to a different url in the same tab

Reply:
The call has various stages. It first gets the AXTree from the main frame, then filters out irrelevant nodes, and so on. Essentially, at any of these stages, it can fail or be in another state, which would require us to pass the callback in various functions. To avoid this, I think we’re using an observer

Additionally, the observer can be called from elsewhere, ensuring that the UI is always up to date.